src/ripple/basics/SlabAllocator.h

```diff
@@ -100,91 +138,109 @@ class SlabAllocator
+    // The size of an item, including the extra bytes requested and
+    // any padding needed for alignment purposes between items:
+    std::size_t const itemSize_;
 
-                if (ret)
-                {
-                    // Use memcpy to avoid unaligned UB
-                    // (will optimize to equivalent code)
-                    std::memcpy(&l_, ret, sizeof(std::uint8_t*));
-                }
-            }
+    // The maximum number of items that this slab can allocate
+    std::size_t const itemCount_;
 
-            return ret;
+    // The size of each individual slab:
+    std::size_t const slabSize_;
+
+    std::uint8_t*
+    allocate(SlabBlock* slab) noexcept
+    {
+        std::lock_guard l(slab->m_);
+
+        std::uint8_t* ret = slab->list_;
+
+        if (ret)
+        {
+            assert(
+                (slab->avail_ != 0) && (slab->avail_ <= itemCount_) &&
+                (ret >= slab->data_) &&
+                (ret + itemSize_ <= slab->data_ + slabSize_));
+
+            // We use memcpy to avoid UB for unaligned accesses
+            std::memcpy(&slab->list_, ret, sizeof(std::uint8_t*));
+            --slab->avail_;
         }
 
-        /** Return an item to this allocator's freelist.
+        return ret;
+    }
+
+    /** Return an item to this allocator's freelist.
 
-            @param ptr The pointer to the chunk of memory being deallocated.
+        @param slab The slab in which to attempt to return the memory
+        @param ptr The pointer to the chunk of memory being deallocated.
+
+        @return true if the item belongs to the given slab and was freed, false
+                     otherwise.
+    */
+    [[nodiscard]] bool
+    deallocate(SlabBlock* slab, std::uint8_t* ptr) noexcept
+    {
+        assert(ptr != nullptr);
 
-            @note This is a dangerous, private interface; the item being
-                  returned should belong to this allocator. Debug builds
-                  will check and assert if this is not the case. Release
-                  builds will not.
-         */
-        void
-        deallocate(std::uint8_t* ptr) noexcept
         {
-            assert(own(ptr));
+            std::lock_guard l(slab->m_);
 
-            std::lock_guard l(m_);
+            if (auto const d = slab->data_; (ptr < d) || (ptr >= d + slabSize_))
+                return false;
 
-            // Use memcpy to avoid unaligned UB
-            // (will optimize to equivalent code)
-            std::memcpy(ptr, &l_, sizeof(std::uint8_t*));
-            l_ = ptr;
-        }
-    };
+            // It looks like this objects belongs to us. Ensure that it
+            // doesn't extend past the end of our buffer:
+            assert(ptr + itemSize_ <= slab->data_ + slabSize_);
 
-private:
-    // A linked list of slabs
-    std::atomic<SlabBlock*> slabs_ = nullptr;
+            // We use memcpy to avoid UB for unaligned accesses
+            std::memcpy(ptr, &slab->list_, sizeof(std::uint8_t*));
```

Comment:
As @seelabs explained in a previous post, this basically creates a linked list, and uses the memory of the block as the storage for the linked list itself.

Let's look at an example: imagine a system where pointers are 4 bytes long, and the slab allocator is configured to allocate 60 bytes at a time, and divvy it up into 3 20 byte blocks.

Here's what this looks like:

```
 +------  slot 0  ------+ +------  slot 1  ------+ +------  slot 2  ------+
[.... .... .... .... ....|.... .... .... .... ....|.... .... .... .... ....|
```

We need to keep track of a "free list" of items. We don't need anything too complicated (we only work with the "head" of the list, nothing too complicated) so a simple singly-linked list will work.

Normally, to keep a linked list we need some kind of "node" structure. Something like:

```
struct Node
{
    Node *next;
};
```

And, if we're going to have a node, we need some memory in which to "store" it.

But remember that we only need a "node" when an item is going to be on the free list, and something is on the free list because there's some memory that is otherwise unused.

So if the size of the unused memory is greater than the size of our node (which, is the size of a pointer!) then we can just use that memory! In other words, we overlay a singly-linked list of free objects onto the memory of those free objects.

Let's imagine that the 60-byte block of memory we have contains one item (slot 1) that's allocated, and two items that are free (slots 0 and 2) . Here's what that looks link:


```
           +------   free   ------+ +------  object  ------+ +------   free   ------+
          [NEXT .... .... .... ....|.... .... .... .... ....|NEXT .... .... .... ....|
          ^   |                                             ^   |
          |   |                                             |   `------> NULL
          |   `---------------------------------------------'
          |
list_ = --+
```

Now, let's imagine that the object that's in use is released. Here's how things change (sorry, the ASCII graphics are getting more complex here):

```
           +------   free   ------+ +------   free   ------+ +------   free   ------+
          [NEXT .... .... .... ....|NEXT .... .... .... ....|NEXT .... .... .... ....|
          ^   |                    ^   |                    ^   |
          |   |                    |   |                    |   `------> NULL
          `----------------------------'                    |
              |                    |                        |
              `--------------------|------------------------'
                                   |
list_ = ---------------------------+
```

Notice how the head linked list now points to slot 1, slot 1 points to slot 0 and slot 0 points to `nullptr` (i.e. the end of the linked list).

Notice too how the "nodes" of the linked list are occupying the same memory that objects do when the slots are allocated.

I hope this helps.

Reply:
Thanks a lot for the detailed explanation, this helps a lot!!

1. Are we using linked lists instead of contiguous memory allocation in the `SlabAllocator` in order to reduce the fragmentation/wastage of memory? I imagine dealing with a contiguous memory blocks would be simpler than with linked lists.

2. In the above example, `itemSize_ == 20 bytes`, `itemCount_ == 3` and `slabSize_ == 60 bytes`. is that correct?

Reply:
Yes, in the example we have `itemSize_ == 20`, `itemCount_ == 3` and `slabSize_ == 60`.

To your first question, we _are_ using a contiguous block of memory in the `SlabAllocator` (in the example above, a 60 byte contiguous block). That's the whole point and it's to reduce fragmentation and the (necessary) overhead that's imposed by general purpose allocators like `malloc` that must be able to hand out chunks of any size.

The linked list is there to allow us to keep track of which "pieces" of the contiguous chunk of memory are in use, and to be able to return them. Remember that deallocations can happen in a different order than allocations.

In the above example, assume the chunk of memory is at address 100. Let's say you try to use the `SlabAllocator` to allocate 3 items.

The memory for the first item will be at address 100 (range: 100-119). The memory for the second item will be at address 120 (range: 120-139). The memory for the third item will be at address 140 (range: 140-159).

Now, assume you are done with the item at address 120 and you return it to the `SlabAllocator`. It needs to know that this item is now available for reallocation, so it keeps track of that in the linked list. It looks like this: `list_` points to address 120. The *value* at address 120 (interpreted as a pointer) points to NULL.

Now, assume you are done with the item at address 140 and you return it to the `SlabAllocator`. It, again, adds the new item to the linked list. It looks like this: `list_` points to address 140. The *value* at address 140 (interpreted as a pointer) points to 120, and the *value* at address 120 points to NULL.

Reply:
thank you, I understand the design 👍 
This does not necessarily pertain to this PR, but I'm thinking about an alternative design choice. 

You could maintain the set of free slots in a `std::unordered_set<int> freeSlabs;` (I'm glossing over the mapping details from integer to the memory address) to achieve the same goals. Allocation is faster because you don't have to traverse through the entire linked list to find the free slab, deallocation is just as fast.

Reply:
okay, that's true. thanks for the explanation 👍